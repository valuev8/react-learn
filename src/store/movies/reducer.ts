import { Movie } from '../../shared/models/movie.type';
import { Reducer } from 'react';
import {
  APPLY_FILTER,
  APPLY_SEARCH,
  APPLY_SORT,
  CREATE_MOVIE,
  DELETE_MOVIE,
  EDIT_MOVIE,
  GET_MOVIES
} from './types';

type Action =
  | { type: typeof GET_MOVIES, payload: Movie[] }
  | { type: typeof CREATE_MOVIE, payload: Movie }
  | { type: typeof EDIT_MOVIE, payload: Movie }
  | { type: typeof DELETE_MOVIE, payload: number }
  | { type: typeof APPLY_FILTER, payload: string }
  | { type: typeof APPLY_SORT, payload: string }
  | { type: typeof APPLY_SEARCH, payload: string }

export type MoviesState = {
  movies: Movie[];
  filteredMovies: Movie[];
  searchQuery: string;
}

const initialState: MoviesState = {
  movies: [],
  filteredMovies: [],
  searchQuery: '',
}

const moviesReducer: Reducer<MoviesState, Action> = (
  state = initialState,
  action
) => {
  switch (action.type) {
    case GET_MOVIES:
      return {
        ...state,
        movies: action.payload,
        filteredMovies: filterByTitle(state.searchQuery, [...action.payload]),
      };

    case CREATE_MOVIE:
      return {
        ...state,
        movies: [action.payload, ...state.movies],
        filteredMovies: [action.payload, ...state.movies],
      };

    case EDIT_MOVIE:
      return {
        ...state,
        movies: editMovie(state.movies, action.payload),
        filteredMovies: editMovie(state.filteredMovies, action.payload),
      };

    case DELETE_MOVIE:
      return {
        ...state,
        movies: deleteMovie(state.movies, action.payload),
        filteredMovies: deleteMovie(state.filteredMovies, action.payload),
      };

    case APPLY_FILTER:
      return {
        ...state,
        filteredMovies: handleFilter(action.payload, state.movies),
      };

    case APPLY_SORT:
      return {
        ...state,
        filteredMovies: handleSort(action.payload, state.filteredMovies),
      };

    case APPLY_SEARCH:
      return {
        ...state,
        searchQuery: action.payload,
        filteredMovies: filterByTitle(action.payload, [...state.movies]),
      };

    default:
      return state;
  }
}

const deleteMovie = (movies: Movie[], movieId: number): Movie[] => {
  return movies.filter((movie: Movie) => movie.id !== movieId)
};

const editMovie = (movies: Movie[], editedMovie: Movie): Movie[] => {
  return movies.map((movie: Movie) => {
    return movie.id === editedMovie.id
      ? { ...editedMovie }
      : { ...movie }
  });
};

const handleFilter = (genreId: string, movies: Movie[]): Movie[] => {
  if (genreId === 'all') {
    return [...movies];
  }

  return [...movies]
    .filter((movie: Movie) => movie.genres.some((genre) => genre.toUpperCase() === genreId.toUpperCase()));
};

const handleSort = (sortKey: string, movies: Movie[]): Movie[] => {
  return [...movies].sort((a, b) => a[sortKey] > b[sortKey] ? 1 : -1);
};

const filterByTitle = (searchQuery: string, movies: Movie[]): Movie[] => {
  return movies.filter((movie) => movie.title.toLowerCase().includes(searchQuery.toLowerCase()))
};

export default moviesReducer;
