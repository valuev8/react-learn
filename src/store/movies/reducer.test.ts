import reducer, { MoviesState } from './reducer';
import { Movie } from '../../shared/models/movie.type';

const initialState: MoviesState = {
  movies: [],
  filteredMovies: [],
  searchQuery: '',
}
const mockMovie = { title: 'test', id: 1 } as Movie;

describe('movies reducer', () => {
  it('should return the initial state', () => {
    // @ts-ignore
    expect(reducer(undefined, {})).toMatchSnapshot()
  })

  it('should handle GET_MOVIES', () => {
    expect(
      reducer(initialState,
        {
          type: 'GET_MOVIES', payload: [mockMovie],
        })
    ).toMatchSnapshot()
  });

  it('should handle CREATE_MOVIE', () => {
    expect(
      reducer(initialState,
        {
          type: 'CREATE_MOVIE', payload: mockMovie,
        })
    ).toMatchSnapshot()
  });

  it('should handle EDIT_MOVIE', () => {
    expect(
      reducer({ ...initialState, movies: [mockMovie] },
        {
          type: 'EDIT_MOVIE', payload: {...mockMovie, title: 'test2'},
        })
    ).toMatchSnapshot()
  });

  it('should handle DELETE_MOVIE', () => {
    expect(
      reducer({ ...initialState, movies: [mockMovie] },
        {
          type: 'DELETE_MOVIE', payload: 1,
        })
    ).toMatchSnapshot()
  });
})
