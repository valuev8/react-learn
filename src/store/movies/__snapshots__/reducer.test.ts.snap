// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`movies reducer should handle CREATE_MOVIE 1`] = `
Object {
  "filteredMovies": Array [
    Object {
      "id": 1,
      "title": "test",
    },
  ],
  "movies": Array [
    Object {
      "id": 1,
      "title": "test",
    },
  ],
  "searchQuery": "",
}
`;

exports[`movies reducer should handle DELETE_MOVIE 1`] = `
Object {
  "filteredMovies": Array [],
  "movies": Array [],
  "searchQuery": "",
}
`;

exports[`movies reducer should handle EDIT_MOVIE 1`] = `
Object {
  "filteredMovies": Array [],
  "movies": Array [
    Object {
      "id": 1,
      "title": "test2",
    },
  ],
  "searchQuery": "",
}
`;

exports[`movies reducer should handle GET_MOVIES 1`] = `
Object {
  "filteredMovies": Array [
    Object {
      "id": 1,
      "title": "test",
    },
  ],
  "movies": Array [
    Object {
      "id": 1,
      "title": "test",
    },
  ],
  "searchQuery": "",
}
`;

exports[`movies reducer should return the initial state 1`] = `
Object {
  "filteredMovies": Array [],
  "movies": Array [],
  "searchQuery": "",
}
`;
